shell/platform/darwin/macos/framework/Source/FlutterMutatorView.mm

```diff
@@ -0,0 +1,361 @@
+
+  // Master clip in global logical coordinates. This is intersection of all clip rectangles
+  // present in mutators.
+  CGRect masterClip = finalBoundingRect;
+
+  self.layer.opacity = 1.0;
+
+  // Gathered pairs of rounded rect in local coordinates + appropriate transform.
+  std::vector<std::pair<FlutterRoundedRect, CGAffineTransform>> roundedRects;
+
+  for (size_t i = 0; i < layer->platform_view->mutations_count; ++i) {
+    auto mutation = layer->platform_view->mutations[i];
+    if (mutation->type == kFlutterPlatformViewMutationTypeTransformation) {
+      transform = CATransform3DConcat(ToCATransform3D(mutation->transformation), transform);
+    } else if (mutation->type == kFlutterPlatformViewMutationTypeClipRect) {
+      CGRect rect = CGRectApplyAffineTransform(FromFlutterRect(mutation->clip_rect),
+                                               CATransform3DGetAffineTransform(transform));
+      masterClip = CGRectIntersection(rect, masterClip);
+    } else if (mutation->type == kFlutterPlatformViewMutationTypeClipRoundedRect) {
+      CGAffineTransform affineTransform = CATransform3DGetAffineTransform(transform);
+      roundedRects.push_back(std::make_pair(mutation->clip_rounded_rect, affineTransform));
+      CGRect rect = CGRectApplyAffineTransform(FromFlutterRect(mutation->clip_rounded_rect.rect),
+                                               affineTransform);
+      masterClip = CGRectIntersection(rect, masterClip);
+    } else if (mutation->type == kFlutterPlatformViewMutationTypeOpacity) {
+      self.layer.opacity *= mutation->opacity;
+    }
+  }
+
+  if (CGRectIsNull(masterClip)) {
+    self.hidden = YES;
+    return;
+  }
+
+  self.hidden = NO;
+
+  /// Paths in global logical coordinates that need to be clipped to.
+  NSMutableArray* paths = [NSMutableArray array];
+
+  for (const auto& r : roundedRects) {
+    CGAffineTransform inverse = CGAffineTransformInvert(r.second);
+    // Transform master clip to clip rect coordinates and check if this view intersects one of the
+    // corners, which means we need to use path clipping.
+    CGRect localMasterClip = CGRectApplyAffineTransform(masterClip, inverse);
+
+    // Only clip to rounded rectangle path if the view intersects some of the round corners. If
+    // not, clipping to masterClip is enough.
+    if (RoundRectCornerIntersects(r.first, ToFlutterRect(localMasterClip))) {
+      CGPathRef path = PathFromRoundedRect(r.first);
+      CGPathRef transformedPath = CGPathCreateCopyByTransformingPath(path, &r.second);
+      [paths addObject:(__bridge id)transformedPath];
+      CGPathRelease(transformedPath);
+      CGPathRelease(path);
+    }
+  }
+
+  // Add / remove path clip views depending on the number of paths.
+
+  while (_pathClipViews.count > paths.count) {
+    NSView* view = _pathClipViews.lastObject;
+    [view removeFromSuperview];
+    [_pathClipViews removeLastObject];
+  }
+
+  NSView* lastView = self;
+
+  for (size_t i = 0; i < paths.count; ++i) {
```

Comment:
iOS had a gesture problem when using this approach: https://github.com/flutter/flutter/issues/46167
I'm not sure how drag/scroll events are handled on mac, it might or might not apply to macOS.

Reply:
I haven't tried to address input yet. It is possible that this won't work with NSGestureRecognizer or rotate/magnify/swipe. Still, I'd want to avoid rasterizing the clip path as much as possible, since it uses additional memory and likely perform quite badly when clip shape changes frequently (i.e. animation).

So if current approach proves problematic and we can't change nesting on fly, alternatively we could always precreate N clip views, where N-1 views would be masked to CAShapeLayer, and the N-th one will contain rest of the paths rasterized.

Reply:
Maybe dragging a PlatformView is a lot less common on macOS.

At least we should test it and document if the issue does exist.

> precreate N clip views, where N-1 views would be masked to CAShapeLayer, and the N-th one will contain rest of the paths rasterized.

This sounds good to me. We can probably also adapt this on iOS to improve performance when there are no complex nested clips.

Reply:
Another thing is that this PR clips to path only when round rect corners are affected. And in future possibly for path-based clip regions, but there's no way to pass those through embedder API yet. So in many cases, when the resulting clip is a rectangle, clipping is performed by container layer bounds set to intersection of clip rectangles. Which should be quite cheap. While I think the iOS version will clip to path even for rectangular clip? It's been a while, I might be misremembering it.

Reply:
> Another thing is that this PR clips to path only when round rect corners are affected.

Nice, this is a good performance enhancement.

> While I think the iOS version will clip to path even for rectangular clip? It's been a while, I might be misremembering it.

Yes you are correct. We can definitely improve iOS performance by doing something similar: setting the frame of the ChildClippingView when the final clip result is a rect, and remove the maskView. This would be a good improvement.